translators/crate.py

```diff
@@ -58,44 +87,52 @@ def _get_isoformat(self, ms_since_epoch):
         return utc.isoformat()
 
 
-    def translate_to_ngsi(self, resultset, keys, table_name):
-        stmt = "select column_name, data_type from information_schema.columns where table_name = '{}'".format(table_name)
+    def translate_to_ngsi(self, resultset, col_names, table_name):
+        """
+        :param resultset: list of query results for one entity_type
+        :param col_names: list of columns affected in the query
+        :param table_name:
+        :return: iterable(NGSI Entity)
+            Iterable over the translated NGSI entities.
+        """
+        stmt = "select entity_attrs from {} where table_name = '{}'".format(METADATA_TABLE_NAME, table_name)
         self.cursor.execute(stmt)
         res = self.cursor.fetchall()
-        table_columns = dict((k, v) for (k, v) in res)
+
+        if len(res) != 1:
+            msg = "Cannot have {} entries in table '{}' for PK '{}'".format(len(res), METADATA_TABLE_NAME, table_name)
+            self.logger.error(msg)
+        entity_attrs = res[0][0]
 
         for r in resultset:
             entity = {}
-            for k, v in zip(keys, r):
-                if k == self.TIME_INDEX_NAME:
+            for k, v in zip(col_names, r):
+                original_name, original_type = entity_attrs[k]
+
+                if original_name in (NGSI_TYPE, NGSI_ID):
+                    entity[original_name] = v
+
+                elif original_name == self.TIME_INDEX_NAME:
                     # TODO: This might not be valid NGSI. Should we include this time_index in the reply?.
```

Comment:
is the time index a data created by crate? in which case, you should not include it in the reply i believe

Reply:
it's created by quantumleap in the table.
In general I agree, it shouldn't be part of the reconstructed entity. Let's see this in the PR of the API with the implementation of the first query method.